crates/exchange-rate-oracle/src/lib.rs

```diff
@@ -260,18 +279,40 @@ pub mod pallet {
 #[cfg_attr(test, mockable)]
 impl<T: Config> Pallet<T> {
     fn begin_block(_height: T::BlockNumber) {
-        if Self::is_max_delay_passed() {
-            Self::report_oracle_offline();
+        if Self::is_invalidated() {
+            RawValuesUpdated::<T>::set(false);
+
+            let mut raw_values: Vec<_> = RawValues::<T>::iter().map(|(_key, value)| value).collect();
+            let min_timestamp = Self::get_current_time().saturating_sub(Self::get_max_delay());
+            raw_values.retain(|value| value.timestamp >= min_timestamp);
+            if raw_values.len() == 0 {
+                Self::report_oracle_offline();
+            } else {
+                let valid_until = raw_values
+                    .iter()
+                    .map(|x| x.timestamp)
+                    .min()
+                    .map(|timestamp| timestamp + Self::get_max_delay())
+                    .unwrap(); // Won't panic as `values` ensured not empty.
```

Comment:
Even if it cannot panic, I don't really like to have `unwrap` in production code. Why not just report the oracle as offline from the error case?